### PR TITLE
[ready] perf: faster reshape

### PR DESCRIPTION
**Pull request by roelofvandijk**
See below.

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/helpers.py                        4               1
tinygrad/shape/shapetracker.py             5               4
------------------------------------------------------------
total                                      9               5
------------------------------------------------------------
lines added in the tinygrad folder: 4
```

**Comment by dc-dc-dc** [comment]
the helpers change looks good, can you make this pr just that and move the sum changes to #1606 since its the same concept

**Comment by geohot** [comment]
Yea the sum thing is sketch, but I'm fine with the helpers change. I wish there was a one liner for it though.

**Comment by roelofvandijk** [comment]
Reworked this a bit.

The logic should be identical, but it's way faster now:

Before: `27781  105.137    0.004  583.905    0.021 shapetracker.py:230(reshape)`
After: `27781   58.578    0.002  139.876    0.005 shapetracker.py:230(reshape)`


Before:
```
codegen         mean runtime:  240.39ms, runs:   245.97,  297.38,  222.51,  224.17,  223.97,  226.17,  227.69,  271.43,  234.77,  229.83
methodcache     mean runtime:  216.14ms, runs:   198.22,  258.19,  199.87,  203.08,  206.22,  206.12,  242.51,  227.72,  211.65,  207.81
profile         mean runtime:  895.53ms, runs:   848.52,  882.69,  935.33,  885.33,  865.64,  888.95,  956.84,  890.45,  896.73,  904.84
```

After:
```
codegen         mean runtime:  187.78ms, runs:   208.70,  254.92,  184.79,  181.37,  173.03,  174.03,  176.59,  175.37,  172.13,  176.82
methodcache     mean runtime:  164.05ms, runs:   148.87,  228.05,  156.17,  158.27,  156.91,  158.15,  157.74,  158.38,  157.94,  159.98
profile         mean runtime:  749.85ms, runs:   728.19,  737.71,  819.47,  768.21,  731.52,  721.25,  760.91,  736.51,  732.48,  762.29
```

**Comment by roelofvandijk** [comment]
Moved the partition change to https://github.com/tinygrad/tinygrad/pull/1625.

**Comment by roelofvandijk** [comment]
@chenyuxyz Could you maybe check if the logic remains the same?

File: tinygrad/shape/shapetracker.py [hunk 1]
```diff
@@ -228,21 +228,21 @@ def expand(self, new_shape: Tuple[Union[Node,int], ...]) -> ShapeTracker:
...
-      new_var, new_val = new_nodes[0], prod(self.shape) // prod(new_ints)
-      if new_var not in self.var_vals:
-        assert new_var.min <= new_val <= new_var.max, f"variable value {new_val} out of range [{new_var.min}, {new_var.max}]"
-        self.var_vals[new_var] = new_val
-      else: assert self.var_vals[new_var] == new_val, f"value conflicts, was {self.var_vals[new_var]}, set to {new_val}"
-    elif not new_nodes: self.var_vals = {}
     if self.views[-1].shape == new_shape: return self
+    if not all(s.__class__ is int for s in new_shape):
+      if all(s.__class__ is int for s in self.shape):
+        new_ints, new_nodes = partition(new_shape, lambda x: x.__class__ is int)
```
Comment (chenyuxyz) [inline tinygrad/shape/shapetracker.py:234]: This is applying a second pass of `__class__` over new_shape in addition to L232. Would the old way with only one pass be faster? Nesting `if all(s.__class__ is int for s in self.shape):` should be fine and faster.

Comment (roelofvandijk) [reply tinygrad/shape/shapetracker.py:235]: This way is faster, the partition call is expensive.

**Pull request closed without merging**